### PR TITLE
gh-107361: strengthen default SSL context flags

### PR DESCRIPTION
**Pull request by woodruffw**
See #107361: this adds `VERIFY_X509_STRICT` to make the default SSL context perform stricter (per RFC 5280) validation, as well as `VERIFY_X509_PARTIAL_CHAIN` to enforce more standards-compliant path-building behavior.

As part of this changeset, I had to tweak `make_ssl_certs.py` slightly to emit 5280-conforming CA certs. This changeset includes the regenerated certificates after that change.

CC @sethmlarson 

* Issue: gh-107361

----
:books: Documentation preview :books:: https://cpython-previews--112389.org.readthedocs.build/

### CODE REVIEW

**Review by gpshead — APPROVED** [review]

File: Doc/library/ssl.rst
Comment (malemburg) [inline Doc/library/ssl.rst:216]: I think you also need to mention these new flags further in the paragraph explaining what the default settings are.

File: Misc/NEWS.d/next/Library/2023-11-24-23-40-00.gh-issue-107361.v54gh46.rst
Comment (malemburg) [inline Misc/NEWS.d/next/Library/2023-11-24-23-40-00.gh-issue-107361.v54gh46.rst:2]: Please also add a mention in the "What's new" document under "Summary – Release highlights / Important deprecations, removals or restrictions:", since this is a potentially breaking behavior change which is rather hidden and will not necessarily surface without direct testing in a production environment (it's well possible that local and CI tests work fine, but when you deploy to a particular internal production environment, things start to fail due to different CA/certificate setups).

**Review by malemburg — CHANGES_REQUESTED** [review]

File: Misc/NEWS.d/next/Library/2023-11-24-23-40-00.gh-issue-107361.v54gh46.rst
Comment (woodruffw) [reply Misc/NEWS.d/next/Library/2023-11-24-23-40-00.gh-issue-107361.v54gh46.rst:2]: Will do, thank you!

**Comment by pitrou** [comment]
Side note: perhaps we mark all PEM files as [generated](https://docs.github.com/en/repositories/working-with-files/managing-files/customizing-how-changed-files-appear-on-github)?

File: Misc/NEWS.d/next/Library/2023-11-24-23-40-00.gh-issue-107361.v54gh46.rst
Comment (woodruffw) [reply Misc/NEWS.d/next/Library/2023-11-24-23-40-00.gh-issue-107361.v54gh46.rst:2]: 2aceb819bed03ba6afedf05bafd7ecd908b0f9e7

File: Doc/library/ssl.rst
Comment (woodruffw) [reply Doc/library/ssl.rst:216]: 15c0313f95fc7f5c73683f81f12669d134b4c5a0

**Comment by woodruffw** [comment]
> Side note: perhaps we mark all PEM files as [generated](https://docs.github.com/en/repositories/working-with-files/managing-files/customizing-how-changed-files-appear-on-github)?

Done: I've marked `certdata/*.{pem,0}` as generated (the `.0` files should also really be `.pem` I think -- I can do some cleanup there in a follow-up PR, if there's interest).

File: Doc/whatsnew/3.13.rst
Comment (vstinner) [inline Doc/whatsnew/3.13.rst:196]: Can you please explain (give instructions) how to get the previous behavior?

Also, how can someone write future-proof code by switching to these defaults in Python 3.12 and older? Maybe also provide instructions for that?

Comment (vstinner) [reply Doc/whatsnew/3.13.rst:196]: Would you mind to explain which kinds of certificates are no longer accepted? If someone is impacted, how certificates should be generated to fit with stricter security policy?

Comment (woodruffw) [reply Doc/whatsnew/3.13.rst:196]: > Can you please explain (give instructions) how to get the previous behavior?

Sure, I can add a note callout similar to the one for SSLv3.

> Also, how can someone write future-proof code by switching to these defaults in Python 3.12 and older? Maybe also provide instructions for that?

The simplest way would probably be to either use `create_default_context()` or `SSLContext()` and twiddle `ctx.verify_flags` directly. 

Do you think it makes sense to include that in the `ssl.rst` docs? It looks like there isn't precedent for doing so (unlike the callout for SSLv3). If so, I'm happy to do it 🙂 

> Would you mind to explain which kinds of certificates are no longer accepted? If someone is impacted, how certificates should be generated to fit with stricter security policy?

This has two parts:

* `VERIFY_X509_PARTIAL_CHAIN` makes chain building strictly _more_ permissive, not less. Specifically, it tells OpenSSL to build chains the way RFC 5280 tells it to, not the bespoke way it does by default. So this flag _alone_ shouldn't cause any certificate rejections, only discoveries of more valid chains 🙂 -- no changes should be required on users' parts.
* `VERIFY_X509_STRICT` tells OpenSSL to validate X.509 certificates per the RFC 5280 profile, which has more extension/criticality requirements than OpenSSL checks by default. In practice, this should have no effect on >99% of users: most root programs use 5280 as their baseline, and the Web PKI's CABF rules are also built on 5280.

TL;DR: No generation changes need to happen for `VERIFY_X509_PARTIAL_CHAIN`. For `VERIFY_X509_STRICT`, the small number of users doing things like `openssl req` manually should ensure that their generated certs conform to RFC 5280 (which they already need to, if they're being used in browsers or other contexts where 5280 or a superset is assumed).

Comment (gpshead) [reply Doc/whatsnew/3.13.rst:196]: > Do you think it makes sense to include that in the `ssl.rst` docs?

Doing it in the what's new makes sense for sure - It doesn't hurt to also include it in the `ssl.rst` versionchanged section. People come at docs from multiple angles depending on what they're trying to do or when and how they are debugging things.

Comment (vstinner) [reply Doc/whatsnew/3.13.rst:196]: I was suggesting to add these information in Python doc. Since tuning a SLContext is complex, IMO it's worth it to elaborate that in ssl.rst and point to that from What's New In Python.

**Comment by vstinner** [comment]
> In practice, this should have no effect on >99% of users: most root programs use 5280 as their baseline, and the Web PKI's CABF rules are also built on 5280.

Your PR required to regenerate all PEM test certificates. Does it mean that they belong to the 1% of affected users?

**Review by sethmlarson — COMMENTED** [review]
I'm ++ on the partial chain changes, my comments below apply to `VERIFY_X509_STRICT`:

Since OpenSSL will likely be the way folks are generating certificates and like you said, OpenSSL allows you to do things that isn't acceptable for Web PKI, is there any value in providing guidance on how one does generate a certificate with OpenSSL that works with `VERIFY_X509_STRICT`?

Also since we're recommending folks down a new code path to restore old behavior, should we create a new test which exercises that code path and verifies that "bad" certificates would be accepted?

What about non-Web PKI, does this change have any affect for certificates in that domain? I am less versed for this, so am relying a bit on others' experience.

**Comment by pitrou** [comment]
> Since OpenSSL will likely be the way folks are generating certificates

Is that still the case? Aren't most users relying on something like Let's Encrypt nowadays (at least for Web certificates, but also perhaps e-mail)?

> is there any value in providing guidance on how one does generate a certificate with OpenSSL that works with `VERIFY_X509_STRICT`?

I don't think writing an OpenSSL command line tutorial is really a responsibility for CPython maintainers.

**Comment by woodruffw** [comment]
> Since OpenSSL will likely be the way folks are generating certificates and like you said, OpenSSL allows you to do things that isn't acceptable for Web PKI, is there any value in providing guidance on how one does generate a certificate with OpenSSL that works with VERIFY_X509_STRICT?

To clarify: I think most users *won't* be using OpenSSL to generate certs -- I think the majority will be getting their certs from a CA service (like Let's Encrypt), with a very small minority doing bespoke things. Putting an OpenSSL example in the docs might cause additional confusion, and IMO won't be too helpful anyways (since the context here is verification, in which case the user might not have any control over the certs they're being presented with).

> What about non-Web PKI, does this change have any affect for certificates in that domain? I am less versed for this, so am relying a bit on others' experience

Nope, this should be wholly compatible with those -- "strict" validation for OpenSSL means "approximate RFC 5280," and the Web PKI profiles are a superset of RFC 5280 🙂

In other words: all public web certificates should be compatible with these changes.

**Comment by malemburg** [comment]
These are the checks that OpenSSL applies when VERIFY_X509_STRICT is set: https://www.openssl.org/docs/man3.0/man1/openssl-verification-options.html#x509_strict (search for -x509_strict in case the link doesn't take you there directly)

These checks all make sense for certs generated with a trusted CA, e.g. Let's Encrypt.

The other common use case is having self-signed certs (e.g. for intranet/VPC devices). Those will already fail to verify with the current standard settings, since self-signed certs are rejected by OpenSSL's verify. You have to use the `ssl._create_unverified_context()` or a custom one to the same effect for those.

Note: The POP/FTP/IMAP and SMTP modules use the unverified context, since intranet/VPC mail and FTP servers will often not use certs created by a trusted CA for various reasons or they are referenced using IP addresses, which don't work with certs at all.

**Review by malemburg — APPROVED** [review]

File: Doc/library/ssl.rst
Comment (vstinner) [inline Doc/library/ssl.rst:158]: I suggest adding instructions on how to get Python 3.12 behavior here. I suppose that these two flags can be cleared with:

```py
context.verify_flags &= ~(_ssl.VERIFY_X509_PARTIAL_CHAIN | _ssl.VERIFY_X509_STRICT)
```

File: Doc/library/ssl.rst
Comment (vstinner) [inline Doc/library/ssl.rst]:
```suggestion
      :data:`VERIFY_X509_STRICT` in its default verify flags.
```

File: Doc/whatsnew/3.13.rst
Comment (vstinner) [reply Doc/whatsnew/3.13.rst:196]: Either link to ssl doc which explains how to get Python 3.12 behavior, or repeat the instructions here. Something like:

```py
context.verify_flags &= ~(_ssl.VERIFY_X509_PARTIAL_CHAIN | _ssl.VERIFY_X509_STRICT)
```

**Review by vstinner — COMMENTED** [review]
I'm in favor of making Python more secure by default, but your PR shows that even Python core devs who maintain PEM files failed to generate certificates properly. So it's likely that at least another user will be affected by stricter default. I'm just asking for clear instructions on how to identify if you are are affected and you're affected, explain how to get the old behavior.

I don't think that Python should explain how to regenerate PEM certificates passing stricter default. I prefer to let users make their own choices.

**Comment by woodruffw** [comment]
Makes sense, I'll make those documentation changes today.

**Comment by malemburg** [comment]
Just to clarify: The CA certs used for testing had to be updated, since the strict check requires these additional constraints:
- The basicConstraints of CA certificates must be marked critical.
- CA certificates must explicitly include the keyUsage extension.

This was not the case for the root CA certs in the test data. After updating the CA certs, all other certs signed with these CA certs had to be resigned.

People will not often create root CA certs by hand (but yes, they might for test cases). So instead of creating uncertainty for regular user certificates, perhaps it's a good idea to highlight these stricter requirements for root CA certs.

File: Doc/library/ssl.rst
Comment (woodruffw) [reply Doc/library/ssl.rst:158]: Will do. FYI: Only `VERIFY_X509_STRICT` would need to be disabled here, since `VERIFY_X509_PARTIAL_CHAIN` doesn't exclude any otherwise valid chains (it only makes OpenSSL include *more* chains that are valid per RFC 5280).

Comment (woodruffw) [reply Doc/library/ssl.rst:158]: 4ae44d1e7b

File: Doc/whatsnew/3.13.rst
Comment (woodruffw) [reply Doc/whatsnew/3.13.rst:196]: Added a callout similar to the one for SSLv3 here: 4ae44d1e7b

That'll get linked via the ref to `:func:ssl.create_default_context` -- do you think another link makes sense, or is that sufficient?

**Comment by sethmlarson** [comment]
@woodruffw I don't see the new test case for ensuring that disabling `X509_VERIFY_STRICT` would allow a "bad" certificate to be used. After that is added I am satisfied with this PR, thanks much!

File: Doc/library/ssl.rst
Comment (vstinner) [inline Doc/library/ssl.rst]:
```suggestion
      may reject pre-:rfc:`5280` or malformed certificates that the
```

File: Doc/whatsnew/3.13.rst
Comment (vstinner) [reply Doc/whatsnew/3.13.rst:196]: Would it be possible to add a link from here to the this new doc? Or copy/paste it here?

File: Lib/ssl.py
Comment (vstinner) [inline Lib/ssl.py:709]: I suggest adding an empty line for readability.

```suggestion
    # first trust anchor, even if that anchor is not self-signed.
    #
```

File: Doc/whatsnew/3.13.rst
Comment (woodruffw) [reply Doc/whatsnew/3.13.rst:196]: I've copied the same `.. note` and example here, PTAL :slightly_smiling_face:

**Comment by woodruffw** [comment]
> @woodruffw I would still like a new test case for ensuring that disabling `X509_VERIFY_STRICT` would allow a "bad" certificate to be used. After that is added I am satisfied with this PR, thanks much!

Making sure I understand: the idea is a "backstop" test to make sure that users have a downgrade path, right? If so, makes sense, and I'll work on that now.

File: Lib/test/test_ssl.py
Comment (woodruffw) [inline Lib/test/test_ssl.py:2962]: @sethmlarson This test provides a backstop check on `VERIFY_X509_STRICT`, PTAL!

**Comment by woodruffw** [comment]
Sorry for the delay here! I've updated this PR to include a testcase for strict verification, including a "backstop" test that ensures that disabling `VERIFY_X509_STRICT` allows this to verify as it did before the change.

**Comment by woodruffw** [comment]
I'm a little mystified by this error, which only happens on OpenSSL 1.1.1w in CI:

```
 0:00:10 load avg: 2.63 [18/43/1] test_ssl failed (1 failure)
test test_ssl failed -- Traceback (most recent call last):
  File "/home/runner/work/cpython/cpython/Lib/test/test_ssl.py", line 2971, in test_verify_strict
    with self.assertRaises(ssl.SSLError):
        s.connect((HOST, server.port))
AssertionError: SSLError not raised
```

AFAICT, the strict flag exists in that version and has the same semantics as in OpenSSL 3.0 and 3.1, which both pass.

**Comment by woodruffw** [comment]
This should be good to go again: I've added a "backstop" test that confirms that `VERIFY_X509_STRICT` enables and disables as expected.

**Review by sethmlarson — APPROVED** [review]
LGTM, apologies for the delay!

**Comment by woodruffw** [comment]
> LGTM, apologies for the delay!

It's "only" 3 months late on my end, so no hard feelings 😉

**Comment by woodruffw** [comment]
Gentle ping on this! I've deconflicted the changelog again.

**Review by gpshead — APPROVED** [review]
LGTM - I'm just running a round of buildbot tests to look for any surprises before merging.

**Comment by sethmlarson** [comment]
@woodruffw Looks like this change breaks trustme generated certificates, see https://github.com/python-trio/trustme/pull/642

**Comment by woodruffw** [comment]
> @woodruffw Looks like this change breaks trustme generated certificates, see https://github.com/python-trio/trustme/pull/642

Thanks for sharing! Looking now.

**Comment by vstinner** [comment]
> @woodruffw Looks like this change breaks trustme generated certificates, see https://github.com/python-trio/trustme/pull/642

If an application is affected by these changes, they can "just" disable the two options, no?

Using `ctx.verify_flags &= ~(ssl.VERIFY_X509_STRICT | ssl.VERIFY_X509_PARTIAL_CHAIN)`:

```py
$ ./python
Python 3.13.0a5+ (heads/main:b44898299a2, Mar 28 2024, 09:01:21) [GCC 13.2.1 20240316 (Red Hat 13.2.1-7)] on linux
>>> import ssl

>>> ctx=ssl.create_default_context()
>>> ctx.verify_flags & ssl.VERIFY_X509_STRICT
<VerifyFlags.VERIFY_X509_STRICT: 32>
>>> ctx.verify_flags & ssl.VERIFY_X509_PARTIAL_CHAIN
<VerifyFlags.VERIFY_X509_PARTIAL_CHAIN: 524288>

>>> ctx.verify_flags &= ~(ssl.VERIFY_X509_STRICT | ssl.VERIFY_X509_PARTIAL_CHAIN)
>>> ctx.verify_flags & ssl.VERIFY_X509_STRICT
<VerifyFlags.VERIFY_DEFAULT: 0>
>>> ctx.verify_flags & ssl.VERIFY_X509_PARTIAL_CHAIN
<VerifyFlags.VERIFY_DEFAULT: 0>
```

**Comment by woodruffw** [comment]
> If an application is affected by these changes, they can "just" disable the two options, no?

Yep, correct. I think the above is also generating certs, so they wanted to get it right more generally, so that each Python downstream consumer doesn't have to also download the strict option.

**Comment by thoughtsunificator** [comment]
It's scary how you can [break people'](https://www.google.com/search?q=CA+cert+does+not+include+key+usage+extension+%22python+3.13%22&sca_esv=7434d0bd66bb8ac8&udm=14&hl=en&sxsrf=AE3TifP2SN1Pl_dhduam8nwwi-zrzPIShw%3A1753537882875&ei=Wt2EaIqVNbuDkdUPuMXByQ4&ved=0ahUKEwiKqu_U1dqOAxW7QaQEHbhiMOkQ4dUDCBA&uact=5&oq=CA+cert+does+not+include+key+usage+extension+%22python+3.13%22&gs_lp=Ehlnd3Mtd2l6LW1vZGVsZXNzLXdlYi1vbmx5IjpDQSBjZXJ0IGRvZXMgbm90IGluY2x1ZGUga2V5IHVzYWdlIGV4dGVuc2lvbiAicHl0aG9uIDMuMTMiMgUQIRigATIFECEYoAEyBRAhGKABMgUQIRigAUicEFDnCljLD3ABeACQAQCYAU2gAbsCqgEBNbgBA8gBAPgBAZgCBqAC7ALCAgYQABgIGB7CAggQABiABBiiBMICBRAAGO8FwgIIEAAYogQYiQWYAwCIBgGSBwE2oAfKErIHATW4B94CwgcHMC4xLjQuMcgHHw&sclient=gws-wiz-modeless-web-only)s stuff and use the 1% nonsense as a rational while getting your own ovation for it.

Instead, what should have been done is providing concrete examples of what DOES break and how to remedy it.

Let's not forget the old story of [certifi](https://github.com/certifi/python-certifi) which decided that it would completely bypass the native store. 

Please stop breaking more stuff.

**Pull request merged**
